src/Microsoft.ML.Data/Model/Onnx/OnnxContext.cs

```diff
@@ -2,245 +2,108 @@
-
-            if (!ContainsColumn(colName))
-                AddVariable(colName);
-
-            colName = GetVariableName(colName);
-            _outputs.Add(OnnxUtils.GetModelArgs(type, colName, dim));
-            return colName;
-        }
-
-        /// <summary>
-        /// Adds an input variable to the list.
-        /// </summary>
-        public void AddInputVariable(ColumnType type, string colName)
-        {
-            _host.CheckValue(type, nameof(type));
-            _host.CheckValue(colName, nameof(colName));
-
-            colName = AddVariable(colName);
-            _inputs.Add(OnnxUtils.GetModelArgs(type, colName));
-        }
+        /// <param name="opType">The name of the ONNX operator to apply</param>
+        /// <param name="inputs">The names of the variables as inputs</param>
+        /// <param name="outputs">The names of the variables to create as outputs,
+        /// which ought to have been something returned from <see cref="AddIntermediateVariable(ColumnType, string, bool)"/></param>
+        /// <param name="name">The name of the operator, which ought to be something returned from <see cref="GetNodeName(string)"/></param>
+        /// <param name="domain">The domain of the ONNX operator, if non-default</param>
+        /// <returns>A node added to the in-progress ONNX graph, that attributes can be set on</returns>
+        public abstract OnnxNode CreateNode(string opType, IEnumerable<string> inputs,
+            IEnumerable<string> outputs, string name, string domain = null);
+    }
 
+    public static class OnnxContextExtensions
+    {
         /// <summary>
-        /// Makes the ONNX model based on the context.
+        /// Convenience alternative to <see cref="OnnxContext.CreateNode(string, IEnumerable{string}, IEnumerable{string}, string, string)"/>
+        /// for the case where there is exactly one input and output.
         /// </summary>
-        public ModelProto MakeModel()
-            => OnnxUtils.MakeModel(_nodes, _producerName, _name, _domain, _producerVersion, _modelVersion, _inputs, _outputs, _intermediateValues);
+        /// <param name="ctx">The ONNX save context</param>
+        /// <param name="opType">The name of the ONNX operator to apply</param>
+        /// <param name="input">The name of the variable as input</param>
+        /// <param name="output">The name of the variable as output,
+        /// which ought to have been something returned from <see cref="OnnxContext.AddIntermediateVariable(ColumnType, string, bool)"/></param>
+        /// <param name="name">The name of the operator, which ought to be something returned from <see cref="OnnxContext.GetNodeName(string)"/></param>
+        /// <param name="domain">The domain of the ONNX operator, if non-default</param>
+        /// <returns>A node added to the in-progress ONNX graph, that attributes can be set on</returns>
+        public static OnnxNode CreateNode(this OnnxContext ctx,
```

Comment:
Do we need this to be an extension method anymore?  Can't it just be a regular method on `OnnxContext`?